2010-10-15  Yao Qi  <yao@codesourcery.com>

	Backport from mainline:

	2010-10-14  Yao Qi  <yao@codesourcery.com>

	gcc/
        PR target/45447
        * config/arm/arm.c (arm_build_builtin_va_list): Assign
        va_list_name to TYPE_STUB_DECL (va_list_type).

	gcc/testsuite/
	PR target/45447
        * gcc.target/arm/pr45447.c: New test.

=== modified file 'gcc/config/arm/arm.c'
Index: gcc-4.5/gcc/config/arm/arm.c
===================================================================
--- gcc-4.5.orig/gcc/config/arm/arm.c
+++ gcc-4.5/gcc/config/arm/arm.c
@@ -1166,6 +1166,7 @@ arm_build_builtin_va_list (void)
 			     va_list_type);
   DECL_ARTIFICIAL (va_list_name) = 1;
   TYPE_NAME (va_list_type) = va_list_name;
+  TYPE_STUB_DECL (va_list_type) = va_list_name;
   /* Create the __ap field.  */
   ap_field = build_decl (BUILTINS_LOCATION,
 			 FIELD_DECL, 
Index: gcc-4.5/gcc/testsuite/gcc.target/arm/pr45447.c
===================================================================
--- /dev/null
+++ gcc-4.5/gcc/testsuite/gcc.target/arm/pr45447.c
@@ -0,0 +1,3 @@
+/* { dg-do compile } */
+/* { dg-options "-g -femit-struct-debug-baseonly" } */
+typedef __builtin_va_list x;
